File: main/components-in-a-graph/components-in-a-graph.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <iterator>
#include <limits>
#include <numeric>
#include <tuple>
#include <utility>
#include <vector>

namespace {
    class DisjointSets {
    public:
        explicit DisjointSets(int count);

        void unite(int i, int j) noexcept;

        std::pair<int, int> minmax_nontrivial_component_sizes() const noexcept;

    private:
        int count() const noexcept;

        int find_set(int i) noexcept;

        void join(int parent, int child) noexcept;

        std::vector<int> parents_;
        std::vector<int> sizes_;
    };

    DisjointSets::DisjointSets(const int count)
        : parents_(count), sizes_(count, 1)
    {
        std::iota(begin(parents_), end(parents_), 0);
    }

    void DisjointSets::unite(int i, int j) noexcept
    {
        // Find the ancestors and stop if they are the same.
        i = find_set(i);
        j = find_set(j);
        if (i == j) return;

        // Union by size.
        if (sizes_[i] < sizes_[j])
            join(j, i);
        else
            join(i, j);
    }

    std::pair<int, int>
    DisjointSets::minmax_nontrivial_component_sizes() const noexcept
    {
        auto min = std::numeric_limits<int>::max();
        auto max = std::numeric_limits<int>::min();

        for (auto i = 0; i < count(); ++i) {
            if (i == parents_[i] && sizes_[i] > 1) {
                min = std::min(min, sizes_[i]);
                max = std::max(max, sizes_[i]);
            }
        }

        return std::make_pair(min, max);
    }

    inline int DisjointSets::count() const noexcept
    {
        return static_cast<int>(parents_.size());
    }

    int DisjointSets::find_set(int i) noexcept
    {
        assert(0 <= i && i < count());

        // Find the ancestor.
        auto j = i;
        while (j != parents_[j]) j = parents_[j];

        // Compress the path.
        while (i != j) {
            const auto parent = parents_[i];
            parents_[i] = j;
            i = parent;
        }

        return j;
    }

    inline void DisjointSets::join(const int parent, const int child) noexcept
    {
        sizes_[parent] += sizes_[child];
        parents_[child] = parent;
    }
}

int main()
{
    std::ios_base::sync_with_stdio(false);

    int n {};
    std::cin >> n;
    DisjointSets sets {n * 2 + 1};

    while (n-- > 0) {
        int i {}, j {};
        std::cin >> i >> j;
        sets.unite(i, j);
    }

    int min {}, max {};
    std::tie(min, max) = sets.minmax_nontrivial_component_sizes();
    std::cout << min << ' ' << max << '\n';
}

File: main/components-in-a-graph/components-in-a-graph.cs
using System;
using System.Collections.Generic;
using System.Linq;

[assembly: System.Reflection.AssemblyVersion("1.0.0.0")]

internal struct MinMax {
    internal MinMax(int min, int max)
    {
        Min = min;
        Max = max;
    }

    internal int Min { get; }

    internal int Max { get; }

    public override string ToString() => $"{Min} {Max}";
}

internal static class EnumerableInt32Extensions {
    internal static MinMax MinMax(this IEnumerable<int> self)
        => self.Aggregate(
                seed: new MinMax(min: int.MaxValue, max: int.MinValue),
                func: (acc, item) => new MinMax(min: Math.Min(acc.Min, item),
                                                max: Math.Max(acc.Max, item)));
}

internal sealed class DisjointSets {
    internal DisjointSets(int count)
    {
        _elems = new int[count];

        // Each element starts in a singleton set (cardinality 1).
        foreach (var i in Enumerable.Range(0, count)) _elems[i] = -1;
    }

    internal void Union(int i, int j)
    {
        // Find the ancestors and stop if they are already the same.
        i = FindSet(i);
        j = FindSet(j);
        if (i == j) return;

        // Unite by size. Sizes are stored as strictly negative values.
        if (_elems[i] > _elems[j]) {
            // The i-set is SMALLER than the j-set. Make j the parent.
            Join(parent: j, child: i);
        } else {
            // The i-set is at least as big as the j-set. Make i the parent.
            Join(parent: i, child: j);
        }
    }

    internal IEnumerable<int> ComponentSizes()
        => from e in _elems
           where e < 0
           select -e;

    private int FindSet(int i)
    {
        // Catches possible errors in client code.
        RangeCheck(nameof(i), i);

        // Find the ancestor.
        var j = i;
        while (_elems[j] >= 0) j = _elems[j];

        // Compress the path.
        while (i != j) {
            var parent = _elems[i];
            _elems[i] = j;
            i = parent;
        }

        return j;
    }

    private void Join(int parent, int child)
    {
        // Catches possible bugs. Might be better as assertions.
        RangeCheck(nameof(parent), parent);
        RangeCheck(nameof(child), child);

        _elems[parent] += _elems[child];
        _elems[child] = parent;
    }

    private void RangeCheck(string name, int i)
    {
        if (i < 0 || i >= _elems.Length)
            throw new ArgumentOutOfRangeException(paramName: name);
    }

    private readonly int[] _elems;
}

internal static class Solution {
    private static MinMax
    MinMaxNontrivialComponentSizes(this DisjointSets self)
        => self.ComponentSizes().Where(count => count > 1).MinMax();

    private static int[] ReadRecord()
    {
        var tokens = Console.ReadLine()
                            .Split((char[])null,
                                   StringSplitOptions.RemoveEmptyEntries);

        return Array.ConvertAll(tokens, int.Parse);
    }

    private static void Main()
    {
        var n = int.Parse(Console.ReadLine());
        var sets = new DisjointSets(n * 2 + 1);

        foreach (var t in Enumerable.Range(0, n)) {
            var record = ReadRecord();
            sets.Union(record[0], record[1]);
        }

        Console.WriteLine(sets.MinMaxNontrivialComponentSizes());
    }
}
